EKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,668 @@
+
+}
+
+void EKFGSF_yaw::initialiseEKFGSF()
+{
+	_gsf_yaw = 0.0f;
+	_ekf_gsf_vel_fuse_started = false;
+	_gsf_yaw_variance = M_PI_2_F * M_PI_2_F;
+	memset(&_ekf_gsf, 0, sizeof(_ekf_gsf));
+	const float yaw_increment = 2.0f * M_PI_F / (float)N_MODELS_EKFGSF;
+	for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index++) {
+		// evenly space initial yaw estimates in the region between +-Pi
+		_ekf_gsf[model_index].X[2] = -M_PI_F + (0.5f * yaw_increment) + ((float)model_index * yaw_increment);
+
+		// All filter models start with the same weight
+		_ekf_gsf[model_index].W = 1.0f / (float)N_MODELS_EKFGSF;
+
+		// take velocity states and corresponding variance from last meaurement
+		_ekf_gsf[model_index].X[0] = _vel_NE(0);
+		_ekf_gsf[model_index].X[1] = _vel_NE(1);
+		_ekf_gsf[model_index].P[0][0] = sq(_vel_accuracy);
+		_ekf_gsf[model_index].P[1][1] = _ekf_gsf[model_index].P[0][0];
+
+		// use half yaw interval for yaw uncertainty
+		_ekf_gsf[model_index].P[2][2] = sq(0.5f * yaw_increment);
+	}
+}
+
+float EKFGSF_yaw::gaussianDensity(const uint8_t model_index) const
+{
+	const float t2 = _ekf_gsf[model_index].S[0][0] * _ekf_gsf[model_index].S[1][1];
+	const float t5 = _ekf_gsf[model_index].S[0][1] * _ekf_gsf[model_index].S[1][0];
+	const float t3 = t2 - t5; // determinant
+	float t4; // determinant inverse
+	if (fabsf(t3) > 1e-6f) {
+		t4 = 1.0f/t3;
+	} else {
+		t4 = 1.0f/t2;
+	}
+
+	// inv(S)
+	float invMat[2][2];
+	invMat[0][0] =   t4 * _ekf_gsf[model_index].S[1][1];
+	invMat[1][1] =   t4 * _ekf_gsf[model_index].S[0][0];
+	invMat[0][1] = - t4 * _ekf_gsf[model_index].S[0][1];
+	invMat[1][0] = - t4 * _ekf_gsf[model_index].S[1][0];
+
+ 	// inv(S) * innovation
+	float tempVec[2];
+	tempVec[0] = invMat[0][0] * _ekf_gsf[model_index].innov[0] + invMat[0][1] * _ekf_gsf[model_index].innov[1];
+	tempVec[1] = invMat[1][0] * _ekf_gsf[model_index].innov[0] + invMat[1][1] * _ekf_gsf[model_index].innov[1];
+
+	// transpose(innovation) * inv(S) * innovation
+	float normDist = tempVec[0] * _ekf_gsf[model_index].innov[0] + tempVec[1] * _ekf_gsf[model_index].innov[1];
+
+	return M_TWOPI_INV * sqrtf(t4) * expf(-0.5f * normDist);
+}
+
+bool EKFGSF_yaw::getLogData(float *yaw_composite, float *yaw_variance, float yaw[N_MODELS_EKFGSF], float innov_VN[N_MODELS_EKFGSF], float innov_VE[N_MODELS_EKFGSF], float weight[N_MODELS_EKFGSF])
+{
+	if (_ekf_gsf_vel_fuse_started) {
+		memcpy(yaw_composite, &_gsf_yaw, sizeof(_gsf_yaw));
+		memcpy(yaw_variance, &_gsf_yaw_variance, sizeof(_gsf_yaw_variance));
+		for (uint8_t model_index = 0; model_index < N_MODELS_EKFGSF; model_index++) {
+			yaw[model_index] = _ekf_gsf[model_index].X[2];
+			innov_VN[model_index] = _ekf_gsf[model_index].innov[0];
+			innov_VE[model_index] = _ekf_gsf[model_index].innov[1];
+			weight[model_index] = _ekf_gsf[model_index].W;
+		}
+		return true;
+	}
+	return false;
+}
+
+void EKFGSF_yaw::makeCovSymEKF(const uint8_t model_index)
```

Comment:
If you make P a `matrix:SquareMatrix(float, 3) you can make it symmetric by calling `P.makeBlockSymmetric<3>(0)`. Then would not need this function.

Reply:
That makes sense. Done.